From: Leo Costela <costela@debian.org>
Date: Sat, 15 Feb 2014 14:34:56 +0100
Subject: use_fontconfig

Use fontconfig if available, instead of a hardcoded font
file. Paves way for splitting the font file into a separate package.

Forwarded: no
---
 Font.cpp     | 40 ++++++++++++++++++++++++++++++++++++----
 Makefile.am  |  6 +++---
 configure.in |  7 +++++++
 3 files changed, 46 insertions(+), 7 deletions(-)

diff --git a/Font.cpp b/Font.cpp
index 3630fd2..fce37ce 100644
--- a/Font.cpp
+++ b/Font.cpp
@@ -18,6 +18,10 @@
 #include "Config.h"
 #include <SDL/SDL_ttf.h>
 
+#ifdef HAVE_FONTCONFIG
+#include <fontconfig/fontconfig.h>
+#endif
+
 #define FONT(fONTpTR) ((TTF_Font*)((fONTpTR)->m_state))
 
 struct FontCanvas : public Canvas
@@ -31,7 +35,34 @@ struct FontCanvas : public Canvas
 Font::Font( const std::string& file, int ptsize )
 {
   TTF_Init();
-  std::string fname = Config::findFile(file);
+#ifdef HAVE_FONTCONFIG
+  std::string fname;
+  if( FcInit() )
+  {
+    FcResult    result;
+    FcChar8 	*fontfile;
+
+    FcPattern   *pat = FcNameParse( (FcChar8 *) file.c_str() );
+    FcConfigSubstitute (0, pat, FcMatchPattern);
+    FcDefaultSubstitute (pat);
+
+    FcPattern   *match = FcFontMatch (0, pat, &result);
+
+    FcPatternDestroy (pat);
+
+    FcPatternGetString(match, FC_FILE, 0, &fontfile);
+
+    fname = std::string((const char*)fontfile);
+  }
+  else
+  {
+    // no fontconfig, fallback to builtin font
+    fname = Config::findFile(file+".ttf");
+  }
+#else
+  std::string fname = Config::findFile(file+".ttf");
+#endif
+
   m_state = TTF_OpenFont( fname.c_str(), ptsize );
   m_height = metrics("M").y;
 }
@@ -93,19 +124,20 @@ void Font::drawWrap( Canvas* canvas, Rect area,
 
 const Font* Font::titleFont()
 {
-  static Font* f = new Font("femkeklaver.ttf",48);
+
+  static Font* f = new Font("femkeklaver",48);
   return f;
 }
 
 const Font* Font::headingFont()
 {
-  static Font* f = new Font("femkeklaver.ttf",32);
+  static Font* f = new Font("femkeklaver",32);
   return f;
 }
 
 const Font* Font::blurbFont()
 {
-  static Font* f = new Font("femkeklaver.ttf",24);
+  static Font* f = new Font("femkeklaver",24);
   return f;
 }
 
diff --git a/Makefile.am b/Makefile.am
index 3135fd2..258dbcb 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -32,8 +32,8 @@ numptyphysics_SOURCES = \
 	OsFreeDesktop.cpp \
 	OsWin32.cpp
 
-numptyphysics_CPPFLAGS = -IXX $(SDL_CFLAGS) $(HILDON_CFLAGS)
-numptyphysics_LDADD = libbox2d.a $(SDL_LIBS) $(HILDON_LIBS)
+numptyphysics_CPPFLAGS = -IXX $(SDL_CFLAGS) $(HILDON_CFLAGS) $(FONTCONFIG_CFLAGS)
+numptyphysics_LDADD = libbox2d.a $(SDL_LIBS) $(HILDON_LIBS) $(FONTCONFIG_LIBS)
 
 numptyphysics_DATA = \
 	data/C00_Title.npz \
@@ -58,7 +58,7 @@ numptyphysics_DATA = \
 	data/close.png \
 	data/record.png \
 	data/theend.png \
-	data/femkeklaver.ttf
+	#data/femkeklaver.ttf
 
 
 libbox2d_a_SOURCES = \
diff --git a/configure.in b/configure.in
index f24740b..da4b50a 100644
--- a/configure.in
+++ b/configure.in
@@ -73,6 +73,13 @@ AC_SUBST(icon_50x50dir)
 AC_SUBST(icon_64x64dir)
 AC_SUBST(icon_scalabledir)
 
+PKG_CHECK_MODULES(FONTCONFIG, fontconfig,
+    [
+        AC_DEFINE([HAVE_FONTCONFIG], 1)
+    ],
+    [
+        AC_MSG_WARN([no fontconfig, using builtin font])
+    ])
 
 AC_CHECK_LIB(stdc++, main)
 AC_CHECK_LIB(X11, main)
